### ISSUE TITLE
improve synced view tracking

### ISSUE THREAD

**Issue by incanus** [mapbox/mapbox-gl-native#1125]
General ticket for improvement of "sticking" native, Cocoa-side views to the panning, zooming GL render view. 

This relates to the user dot now, later to callouts and annotation views. Let's keep a ticket around until we are 100% satisfied with the behavior here, but we can move it along milestones as we make staged improvements.

**Comment by incanus** [comment]
I'm feeling pretty good about the current state of tracking, especially after the above commit which avoids invalid `CLLocationCoordinate2D`-based updates during `TransformState` "will change" events (and just catches the "did change"). 

The one issue I'm seeing which I'd like to clean up is the dot disappearing during e.g. a double-tap animated zoom. It doesn't show up again until the next `CLLocationManager` update. It gets hidden because the `CGPoint` received out of `-[MGLMapView convertCoordinate:toPointToView:]` is well out of range of given screen bounds. I think this is just some sort of math error in `TransformState` during animations.

**Comment by incanus** [comment]
Another lead that confirms the above: pan the map until the dot is >150px off screen, then hit the user locate button. The map pans back into view (meaning the map centering to the valid user location is of course working) but the dot doesn't show up until the next CL update post-animation. During the animation it's still residing at the offscreen value, so it's hidden.

**Comment by incanus** [comment]
Tracking is pretty good right now. Only outstanding issue I'd like to hit for the beta is the dot disappearing briefly during animated zooms (but coming back at least, now). I'm nearing some completion on this but don't want to hold up a build tonight so I'll work on that a bit over the weekend. 

Going to squash & merge a few commits now with improvements thus far.

**Comment by incanus** [comment]
First batch of improvements in https://github.com/mapbox/mapbox-gl-native/pull/1151.

**Comment by incanus** [comment]
Main blocker here now captured in the more general https://github.com/mapbox/mapbox-gl-native/issues/1160 bug with projection conversion.

**Comment by kkaefer** [comment]
FWIW, I think we should get rid of overlaying iOS views onto the Map and instead render everything ourselves. Apple recommends against doing this as well: https://developer.apple.com/library/ios/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/Performance/Performance.html#//apple_ref/doc/uid/TP40008793-CH105-SW8

**Comment by incanus** [comment]
Right, but I direct your attention back to https://github.com/incanus/AniAnnoDemo and what needs to happen to make this work :-) The thing I'm most uncertain about is syncing frames of animation from native views into GL. Maybe it's not as terrible as I'm imagining though.

**Comment by mb12** [comment]
@incanus Perhaps, we should try posting the synchronization problem to Apple developer forums and/or use the 3  incidents that Apple includes in developer support. Apple is already doing this for Mapkit. Its a problem that they have already solved. It would make writing all kind of Map overlays super easy.

https://developer.apple.com/devforums/

**Comment by incanus** [comment]
http://stackoverflow.com/a/17096916/977220

> In fact, you can't synchronize them using current APIs. MobileMaps.app and Apple Map Kit use private property on `CAEAGLLayer` `synchronized` to workaround this issue.

I'm intending to dig deep when the time comes for this, but right now the user dot is syncing well and we just need the bug fix in https://github.com/mapbox/mapbox-gl-native/issues/1160 for a workable version right now. 

From the POV of annotations, after shapes, this will be my next priority.

**Comment by incanus** [comment]
On this front, I've been thinking about looking into routes for capture a view's layer's `presentationLayer` at a given point in time and rasterizing it to a GL render layer. That is, all tracked views atop the GL view will actually get rasterized into a single "layer" in the GL. I have no idea how feasible this is, but is the sort of approach I've been mulling over researching.

**Comment by incanus** [comment]
We are ok here for b2 and b3 and beyond will feature much more robust point annotations.

**Comment by friedbunny** [comment]
Per WWDC and chat and [Stack Overflow](http://stackoverflow.com/a/30722276/2094275):

> In iOS 9 `CAEAGLLayer` has `presentsWithTransaction` property that synchronizes the two.

**Comment by incanus** [comment]
As part of #1784, I'm looking into this again. While the iOS 9 API looks intriguing, we need a solid iOS 7 & 8 approach anyway. 

Using what's in [`view-based-points`](https://github.com/mapbox/mapbox-gl-native/compare/view-based-points) works fairly well, however some lag is evident. Using the Time Profiler in `Release` configuration reveals our three main resource hogs: 

1. `-[UIView setCenter:]` at 7.5%. This is called across each annotation view to update its position every frame sync. 

![screen shot 2015-06-24 at 12 10 48 pm](https://cloud.githubusercontent.com/assets/17722/8339169/796ff16e-1a6a-11e5-91b7-03e1b221ea44.png)

1. `CALayer` animation commit internals at 2.1%. This improves the smoothness of syncing at a slight cost. 

![screen shot 2015-06-24 at 12 11 13 pm](https://cloud.githubusercontent.com/assets/17722/8339182/98ed7ae8-1a6a-11e5-97d8-60998458be11.png)

1. `GLKView` internal GL rendering at 1.9%. 

![screen shot 2015-06-24 at 12 10 59 pm](https://cloud.githubusercontent.com/assets/17722/8339194/a9ceda32-1a6a-11e5-935d-8eb261f896cc.png)

Next steps are to learn from similar approaches for views/layers in our raster SDK and explore: 

- `CAScrollLayer` which updates `-scrollToPoint:` during pans/rotates, which is where the lag is most noticeable, reframing itself on zooms. This collects all pan updates into one system-maintained position update, effectively, by using a class optimized for collective panning. 

- Possible `UIScrollView` for a similar approach, keeping in mind the most annoying limitation to ever emerge from our raster SDK, https://github.com/mapbox/mapbox-ios-sdk/issues/72.

**Comment by incanus** [comment]
Tried a number of things in combination with reducing the number of `UIView` syncing calls (https://github.com/mapbox/mapbox-gl-native/commit/6987536957f5e5a3d8bf75307a6834ab891a6be6) such as: 

- Directly driving `-[GLKView display]` instead of relying on the normal `UIView`-like `-setNeedsDisplay` and subsequent `-[GLKViewDelegate glkView:drawInRect:]` trigger, which only serves to move both GL and native-side views the same interval and doesn't change their relative relationship. 

- Various combination of `UIView` and `CATransaction` animations wrapping both the GL render call and the `UIView` updates, which doesn't improve things noticeably due to the asynchronous nature that's allowed to be configured in the new iOS 9 API. This includes experimenting with the current private API `-[EAGLLayer asynchronous]` property (http://openradar.appspot.com/radar?id=3118401). 

- Driving `UIView` updates in a new `-[UIView drawRect:]`, which also suffers the same lag due to the nature of `-setNeedsDisplay`. 

- Experimenting with calling GL renders from a non-main queue. 

    - http://codereply.com/answer/4p0288/cadisplaylink-opengl-rendering-breaks-uiscrollview-behaviour.html

Now going to try a perhaps-crazy approach of: 

 - Tracking `UIView` annotations but `hidden`. 
 - Rasterizing the views' layers' `presentationLayer` throughout possible animation updates. 
 - Using the rasterized copy as a texture on a quad in the GL view. 

Refs: 

- http://stackoverflow.com/questions/10120962/combining-uiview-animation-blocks-and-opengl-es-rendering
- http://gamesfromwithin.com/wp-content/uploads/2010/03/GDC10_uikit_opengl.pdf
- http://www.tagwith.com/question_466813_how-to-track-a-zooming-uiscrollview
- http://stackoverflow.com/questions/12772723/replicating-uiview-drawrect-in-opengl-es?rq=1

Might be prohibitive, but if we could avoid constantly uploading the vertices, re-uploading unchanged textures, and doing anything with non-viewport relevant annotations, it just might work. We could also eventually improve this to handle alpha blended annotation views, intermingled z-ordering of native views and GL symbols, and the like. 

Will start with a single annotation first to prove the concept.

**Comment by mb12** [comment]
Does the issue get fixed on iOS 9 beta with presentsWithTransaction?

**Comment by incanus** [comment]
The good news: I got a quad billboard drawing that updates its position based on the associated annotation. No texture yet, just a dumb colored quad. 

The bad news: It still has jitter, just like a native `UIView`. So even drawing in GL atop the map view post-`renderSync()` isn't enough to prevent this. Sort of makes sense; Core Animation is also compositing at this step, so same adverse result. 

The next thing I'm going to try is setting up a mechanism for communicating between GL and iOS about which annotations have iOS-side textures that will be provided (i.e. rasterized `CALayer` objects), changing those from normal `Symbol` rendering to a special quad, and communicating over the boundary what the actual texture is upon `CALayer` update iOS-side, uploading the texture to the GPU and applying it to the quad.

**Comment by incanus** [comment]
I think this meshes well with the MapKit-style annotation view reuse method. 

1. Rasterize a view's layer and upload it as a texture, making a note of the time. 
1. As long as the view/layer doesn't change appearance, don't re-upload the texture, even when instantiating new, reused annotation views with the same identifier. 
1. Upon view/layer change for each annotation, update the recorded time, then re-upload the texture from a new rasterized snapshot as needed. This forks the number of textures across views with identical identifiers, but only when absolutely necessary. As long as the views aren't visibly changing, the same single texture can be reused across multiple annotations.

**Comment by incanus** [comment]
Encapsulating both the GL render calls and the `UIView`-based updates in a `CATransaction` block while using `[(CAEAGLLayer *)self.glView.layer setPresentsWithTransaction:YES]` seems to have no effect. Trying to figure out if I'm using it right.

**Comment by incanus** [comment]
I will push some code shortly to demo, but I am seeing some odd behavior. 

* If in the `GLKView` draw routine, I call `renderSync` to do the normal map draw, then call my own GL commands (with a small, dumb shader) to draw a quad immediately afterwards (in the main thread), the quad jitters and lags a bit in comparison to map pans and zooms. 

* If instead I take the same quad drawing code and put it into `MapContext::render` to call immediately after `painter->render`, there is no lag and the quad is perfectly in sync with the map. In this case, I am calling my custom GL code on the map thread. 

Since `renderSync` is synchronous, I would think that in the first case it wouldn't matter that I'm drawing from two different threads, but that approach may be suspect. 

Note that in the long run, I wouldn't be doing this sort of work on the main thread anyway so as to not impact UI responsiveness, so maybe I should just skip to an approach that draws the quad in the latter way anyway. I can next work on getting `CALayer` snapshots across the wire to use as textures for the quads.

**Comment by incanus** [comment]
It is also a possibility that `GLKView` is doing something smart around automatic buffer swapping based on calling thread.

**Comment by incanus** [comment]
A method to compare the two approaches: 

* Jitter: Grab b02dc95b798a89d1f1a6ccc18329eb4ad1bdad6e, zoom the viewport to DC, add "100 annotations" (which will really be just one), and see a green quad on the map. Pan and zoom and observe jitter. 

* No jitter: Grab bf98dfa5659ce489ef99333c468a01272f8dc468, zoom the viewport to DC, and see the [hardcoded](https://github.com/mapbox/mapbox-gl-native/commit/bf98dfa5659ce489ef99333c468a01272f8dc468#diff-2e6ee6e545f1af2b91a4658698ef2638R336), automatically-added quad. It tracks the map perfectly.

**Comment by incanus** [comment]
I've confirmed that `-[EAGLContext presentRenderbuffer]` is only ever called by `GLKView` internals from the main thread, and is thus not called between the call to `renderSync` and any external drawing ("`end gl marker`" here) is done. I did this by subclassing `EAGLContext` to log the method. 

```
2015-06-29 11:17:59.567 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.570 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.582 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.583 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.583 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.588 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.599 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.600 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.600 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.604 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.615 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.617 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.617 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.621 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.631 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.633 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.633 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.655 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.666 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.667 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.667 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.670 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.682 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.683 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.683 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.688 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.699 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.700 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.700 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:17:59.703 Mapbox GL[349:17692] start renderSync
2015-06-29 11:17:59.713 Mapbox GL[349:17692] end renderSync
2015-06-29 11:17:59.714 Mapbox GL[349:17692] end gl marker
2015-06-29 11:17:59.715 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:18:00.721 Mapbox GL[349:17692] start renderSync
2015-06-29 11:18:00.737 Mapbox GL[349:17692] end renderSync
2015-06-29 11:18:00.738 Mapbox GL[349:17692] end gl marker
2015-06-29 11:18:00.738 Mapbox GL[349:17692] presentRenderBuffer
2015-06-29 11:18:01.350 Mapbox GL[349:17692] start renderSync
2015-06-29 11:18:01.367 Mapbox GL[349:17692] end renderSync
2015-06-29 11:18:01.368 Mapbox GL[349:17692] end gl marker
2015-06-29 11:18:01.368 Mapbox GL[349:17692] presentRenderBuffer
```

**Comment by incanus** [comment]
Next up is comparing the `TransformState` between the use inside of `renderSync` and outside as provided by the getter. Perhaps the geography/pixel conversion routines differ.

**Comment by incanus** [comment]
Transform is not a factor as the transformed screen point for a given annotation is always the same. Checking next the height, width, and other factors that might vary (for some reason?) between frames.

**Comment by incanus** [comment]
Upload vertex positions are the same between core-side and iOS-side GL markers, but the lag persists: 

![anigif-1435595751](https://cloud.githubusercontent.com/assets/17722/8412998/7036978e-1e5b-11e5-8dfb-77449d359c26.gif)

_Black square is drawn inside of `renderSync`, green square is drawn in OpenGL immediately afterwards on the iOS side_ 

I also checked that the vertex shaders were using the same basic procedure, as well as precision, and they are — just a simple: 

```glsl
attribute vec2 a_pos;
void main() {
   gl_Position = vec4(a_pos, 0.0, 1.0);
};
```

The core-side shader applies `u_matrix`, but in the case of these annotations, we pass the `painter->identifyMatrix` since we are doing the screen conversion ourselves. 

-

One thing that does jump out at me in the above GIF is that finger-down drags seems to sync perfectly, but coasting/decelerated pans are where things get out of sync. Looking into that now.

**Comment by jfirebaugh** [comment]
> One thing that does jump out at me in the above GIF is that finger-down drags seems to sync perfectly, but coasting/decelerated pans are where things get out of sync. Looking into that now.

That sounds telling. Please rebase on master and put the main-thread drawing [here](https://github.com/mapbox/mapbox-gl-native/blob/f675d233142bbbfbe77c0145128c3eedfb18e824/src/mbgl/map/map.cpp#L50).

The `Transform` update for easing happens on the next line. In the code where your branch is currently, it happens on the rendering thread, so your code is racing against it.

**Comment by incanus** [comment]
:+1: Was closing out some hanging code from before the weekend; rebasing now to get up to speed.

**Comment by incanus** [comment]
> Please rebase on master and put the main-thread drawing here.

FWIW this requires making the `MapContext`'s `painter` public to `Map` so that we can use the shaders.

**Comment by incanus** [comment]
Disregard last comment. Just merged `master`. 

Black square was custom GL drawing in `MapContext`, as a comparison with...

Green square, which was custom GL drawing in `MGLMapView` after calling `Map::renderSync`, which lagged. 

After applying `master`, I just tried only the green square and the lag is gone. This is because we now drive rendering from the main thread, then do this custom GL code also in the main thread. 

This should be enough for me to next: 

1. Test straight `UIView` / `CATransaction`, but I'm guessing we'll still see the problem that's purportedly solved by the new iOS 9 API. 
1. Regardless, move on to snapshotting `CALayer` and uploading as a texture in the `MGLMapView` custom GL code.

**Comment by incanus** [comment]
`UIView` tracking directly on the map view now works without jitter! Both the user dot and custom annotation views returned by the delegate track nicely. 

Going to double-check that I'm not doing anything in particular in my branch by also testing `master` with the user dot, but if not, we can close here as solved by https://github.com/mapbox/mapbox-gl-native/pull/1624 and move back to building out the annotation view API in https://github.com/mapbox/mapbox-gl-native/issues/1784.

**Comment by incanus** [comment]
The user dot in `master` does behave smoothly and without jitter, but in both branches, I do notice a lag / movement of supposedly static markers during fast panning. I made sure that subtle user location changes weren't at fault by testing on my branch with a custom annotation view with a fixed coordinate. 

Here's a device video capture:  https://dl.dropboxusercontent.com/u/575564/floater.mov

Note how the fixed marker has varying locations, and you can see the subtle floating during a slo-mo playback of the video. 

![screen shot 2015-06-29 at 3 41 10 pm](https://cloud.githubusercontent.com/assets/17722/8416790/13a46b0c-1e76-11e5-9281-7289cabda94e.png)

_initial location_ 

![screen shot 2015-06-29 at 3 40 59 pm](https://cloud.githubusercontent.com/assets/17722/8416795/29e5703c-1e76-11e5-97c9-bdc109463fd2.png)

_frame capture during fast panning_ 

![screen shot 2015-06-29 at 3 41 05 pm](https://cloud.githubusercontent.com/assets/17722/8416836/6b61e9dc-1e76-11e5-950e-bb35f334613e.png)

_another frame capture_ 

So it's not just perception; it's really moving slightly. **This** may be what the OpenGL / `CATransaction` sync-up is about, in which case, we _may_ still want GL-side snapshotting. 

More understanding / definitely a better effect here after #1624, but...

![](http://3.bp.blogspot.com/-v1VpappglbI/TmqrCF7SJpI/AAAAAAAAACE/G6_MFRDFefA/s1600/guy_back_to_square_1_hg_wht.gif)

**Comment by incanus** [comment]
The drift happens in main thread GL drawing code as well as just `UIView` syncing, so it's not the `CATransaction` thing. 

We could live with the drifting if we had to for now, I think, but I'll keep digging on it a bit.

**Comment by incanus** [comment]
Procedure for replicating / testing the "green square hack" — aGL quad drawn in the main thread immediately after `renderSync` and inside of `MGLMapView`: 

1. Grab 1fb66068f35d5084315fdeb493191d5814773076. 
1. Center so that DC is visible. 
1. Add "100" annotations (adds just a single for now). 
1. Observe green quad lag (but no jitter).

**Comment by incanus** [comment]
I can pretty much again isolate lag (and what used to be jitter) to the pan deceleration / animation routine upon end of a pan gesture in iOS. Keeping you finger on the screen prevents the lag from appearing.

**Comment by incanus** [comment]
Currently trying to isolate whether it's specific to `Transform` or specific to gestures by writing up some animated transition code that tries to replicate the lag without gestures being involved (so, still eventtually calling `startTransition`).

**Comment by incanus** [comment]
Using a patch like https://gist.github.com/incanus/41d85a875b032558a6ec proves that it's related to animated transitions and is not specific to gestures.

**Comment by incanus** [comment]
@jfirebaugh nailed this exactly in chat: 

> renderSync does two things in this order: 1. render, 2. update transition for easing
> if you draw after both of those, it will be out of sync
> you need to draw in between those two things

This is why the black square demo worked fine — that's exactly where I placed the draw code. 

Refactoring things into a split between rendering and updating transitions lets this work. The green square, drawn in iOS code, doesn't lag! 

Working on cleaning that up.

**Comment by mb12** [comment]
Does this explain the difference in syncing behavior between downward finger drag and upward finger drag?

**Comment by incanus** [comment]
I’m not familiar with that @mb12 — can you elaborate?

**Comment by mb12** [comment]
I was referring to this comment on the gif animation. 

https://github.com/mapbox/mapbox-gl-native/issues/1125#issuecomment-116754125

**Comment by incanus** [comment]
Breaking things into two parts also fixes syncing issues with straight-up `UIView` annotations. 

@mb12 Yes, there is now no lag.

**Comment by incanus** [comment]
While this will certainly be one of the weirder animations posted here, bear with me a moment. 

I have a prototype of the ability to sample an animating `UIView` and upload a snapshot of it to a texture to be rendered by OpenGL each frame. This points at the future possibility of truly using reuse identifiers for annotation views and uploading them as reusable textures. It remains to be seen whether this is a performance gain, but I'm guessing that it could have benefits over simply tracking `UIView` objects on top of the GL view directly, plus we could reuse annotations views (animating or not) to move things from duplication in the view hierarchy to a memory cost savings in GL textures. 

![anigif-1435732271](https://cloud.githubusercontent.com/assets/17722/8448798/58ca1e56-1f99-11e5-9a14-94b225bbfc37.gif)

_(Full-res video capture of this [here](https://dl.dropboxusercontent.com/u/575564/presolayer.mov))_

What is happening here is the following: 

1. There is a normal `UIView` (center right) with an image as its `layer.contents` which has a repeating rotation animation applied to it. Consider this a potential annotation view created in the expected way. 
1. When frames are being redrawn, i.e. when the map is being moved, every frame we are sampling the `view.layer.presentationLayer.contents` into a `CGImage`. 
1. We are also sampling the `view.layer.presentationLayer.transform`. 
1. We use the image and the transform to draw into a `CGContext`, capturing the raw pixel array, which can then be uploaded as an OpenGL texture. 
1. That texture is drawn on a simple quad (the weird tracking image). 

In practice, the actual `view` would be `hidden` so that we only see the OpenGL one, incurring no resource usage to render the original `UIView`. 

The weird in-and-out-of-frame bit is because the animation rotation origin and the context transform point of reference aren't the same and could be figured out with a little trial and error, causing the texture to appear to rotate in place just like the native view instead of rotate in and out of the quad. Note that the rotation is backwards, too. But the general proof is there that a `UIView` can be sampled as a texture into GL frames, smoothly animating. 

-

I think we should punt on animation annotation views for now, as well as this texture upload business, but it's promising for future use. 

I'm next going to focus on just a bare-bones annotation view API (no actual reuse identifiers), which I should be able to wrap pretty easily now that we've solved the syncing issues.

**Comment by incanus** [comment]
We could further optimize by passing the `view.layer.presentationLayer.transform` into the shader as the projection matrix, avoiding the need to re-upload the texture during animations. Then we'd only have to sample the view to a texture after `view.layer.contents` changes (which should be KVO observable). That would also let us reuse the same texture for all views with the same reuse identifier, since the matrix would vary for each quad being drawn (depending on animation state) but the texture wouldn't.

**Comment by incanus** [comment]
Going the basic `UIView` route for now in #1784, but first we need to split the `renderSync` call per https://github.com/mapbox/mapbox-gl-native/pull/1813.

**Comment by incanus** [comment]
Per chat, @kkaefer is going to look into this & https://github.com/mapbox/mapbox-gl-native/pull/1813 to improve our current user dot.

**Comment by incanus** [comment]
I'm grabbing this for the b3 release branch.

**Comment by incanus** [comment]
We are good for `release-0.5.0` now as of https://github.com/mapbox/mapbox-gl-native/commit/b8388168dd130c67c77254565cdb576df7905915 and https://github.com/mapbox/mapbox-gl-native/commit/a67e80219f90c409dfce8bfef91d9131dabd8d4f; moving off of milestone.

**Comment by ljbade** [comment]
Making note of #2384 as it looks like iOS 9 has some fixes that can be used.

**Comment by incanus** [comment]
We started looking at that in https://github.com/mapbox/mapbox-gl-native/issues/1125#issuecomment-111636947 @ljbade. No help currently.

**Comment by 1ec5** [comment]
#3683 finally synchronizes the user dot with the map. Keeping this issue open until we’re able to leave callouts open while the viewport moves.

**Comment by 1ec5** [comment]
#3683 has landed and will be in v3.2.0. Closing in favor of #3154.
